gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,510 @@
+  // This remains unchanged of position(long) method call.
+  private long contentChannelPosition = -1;
+  private long contentChannelEnd = -1;
+
+  // Used as scratch space when reading bytes just to discard them when trying to perform small
+  // in-place seeks.
+  private byte[] skipBuffer = null;
+  // Size of buffer to allocate for skipping bytes in-place when performing in-place seeks.
+
+  // Prefetched footer content.
+  private byte[] footerContent;
+
+  public GoogleCloudStorageClientReadChannel(
+      @Nonnull Storage gcs,
+      @Nonnull GoogleCloudStorageItemInfo itemInfo,
+      @Nonnull GoogleCloudStorageReadOptions readOptions)
+      throws IOException {
+    validate(itemInfo);
+    this.storage = gcs;
+    this.resourceId =
+        new StorageResourceId(
+            itemInfo.getBucketName(), itemInfo.getObjectName(), itemInfo.getContentGeneration());
+    this.blobId =
+        BlobId.of(
+            resourceId.getBucketName(), resourceId.getObjectName(), resourceId.getGenerationId());
+    this.readOptions = readOptions;
+    this.objectSize = itemInfo.getSize();
+    this.randomAccess = readOptions.getFadvise().equals(Fadvise.RANDOM) ? true : false;
+    this.footerSize = (int) readOptions.getMinRangeRequestSize();
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    throwIfNotOpen();
+
+    // Don't try to read if the buffer has no space.
+    if (dst.remaining() == 0) {
+      return 0;
+    }
+
+    logger.atFiner().log(
+        "Reading %d bytes at %d position from '%s'", dst.remaining(), currentPosition, resourceId);
+
+    if (currentPosition == objectSize) {
+      return -1;
+    }
+
+    int totalBytesRead = 0;
+    while (dst.hasRemaining()) {
+      try {
+        performPendingSeeks(dst.remaining());
+        int bytesRead = contentReadChannel.read(dst);
+
+        if (bytesRead > 0) {
+          totalBytesRead += bytesRead;
+          currentPosition += bytesRead;
+          contentChannelPosition += bytesRead;
+        }
+
+        if (bytesRead < 0) {
+          if (currentPosition != contentChannelEnd && currentPosition != objectSize) {
+            throw new IOException(
+                String.format(
+                    "Received end of stream result before all requestedBytes were received;"
+                        + "EndOf stream signal received at offset: %d where as stream was suppose to end at: %d for resource: %s of size: %d",
+                    currentPosition, contentChannelEnd, resourceId, objectSize));
+          }
+          // If we have reached an end of a contentChannel but not an end of an object.
+          // then close contentChannel and continue reading an object if necessary.
+          if (contentChannelEnd != objectSize && currentPosition == contentChannelEnd) {
+            closeContentChannel();
+            continue;
+          } else {
+            break;
+          }
+        }
+      } catch (Exception e) {
```

Comment:
Not something which we do in [GoogleCloudStorageReadChannel](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/branch-2.2.x/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageReadChannel.java#L388) . Caller can just retry with same `bytebuffer` as we have updated the currentPosition even in case of partialRead before exception was thrown.

Reply:
I see you are now subtracting the partial reads?

Reply:
not `subtracting` but `adding` to currentPosition.